In [1]:
import requests

# 0. Install requests

You can find the official doc [here](https://docs.python-requests.org/en/latest/user/install/#install)

```bash
# -m means run pip module as a script, to avoid the pip.py
python -m pip install requests
```

# 1. Make a request

Now, let’s try to get a webpage. For this example, let’s get GitHub’s public timeline:

In [2]:
response = requests.get("https://api.github.com/events")

All the information that server sends back to us are in the response. In below example, we print the response code 200, which means success. We will show details of response in section 2

In [3]:
print(response)

<Response [200]>


# 1.1 Other http request

Off course, requests can do other possible http request type, below are some code example:

```python
requests.post('https://httpbin.org/post', data={'key':'value'})
requests.put('https://httpbin.org/put', data={'key':'value'})
requests.delete('https://httpbin.org/delete')
requests.head('https://httpbin.org/get')
requests.patch('https://httpbin.org/patch', data={'key':'value'})
requests.options('https://httpbin.org/get')
```
Note for put, post, patch, the data parameter allows you to pass request payload.

# 1.2 Params for get

When you make a quest for http, you can also pass parameter in the URL. below is an example

In [9]:
get_params = {'key1': 'value1', 'key2': 'value2'}
r1 = requests.get('https://httpbin.org/get', params=get_params)

In [10]:
# you can check the url of your request by using .url
print(r1.url)

https://httpbin.org/get?key1=value1&key2=value2


# 2 Response content

We know how to make a request now, Now let's check the response of the request

In [11]:
print(response.text)

[{"id":"20893240420","type":"PullRequestReviewEvent","actor":{"id":29139814,"login":"GCsabi","display_login":"GCsabi","gravatar_id":"","url":"https://api.github.com/users/GCsabi","avatar_url":"https://avatars.githubusercontent.com/u/29139814?"},"repo":{"id":466697561,"name":"joladnijo/joladnijo-webapp","url":"https://api.github.com/repos/joladnijo/joladnijo-webapp"},"payload":{"action":"created","review":{"id":918715493,"node_id":"PRR_kwDOG9E9Wc42wnxl","user":{"login":"GCsabi","id":29139814,"node_id":"MDQ6VXNlcjI5MTM5ODE0","avatar_url":"https://avatars.githubusercontent.com/u/29139814?v=4","gravatar_id":"","url":"https://api.github.com/users/GCsabi","html_url":"https://github.com/GCsabi","followers_url":"https://api.github.com/users/GCsabi/followers","following_url":"https://api.github.com/users/GCsabi/following{/other_user}","gists_url":"https://api.github.com/users/GCsabi/gists{/gist_id}","starred_url":"https://api.github.com/users/GCsabi/starred{/owner}{/repo}","subscriptions_url":"

You can notice the github server send back the event timeline.

Note: Requests will automatically decode content from the server. Most unicode charsets are seamlessly decoded.

When you make a request, Requests makes educated guesses about the encoding of the response based on the HTTP headers. The **text encoding guessed by Requests is used when you access r.text**. You can find out what encoding Requests is using, and change it, using the **.encoding** property.

The encoding guessed by Requests is utf-8, we can change it by assigning a new value

## 2.1 Encoding

In [13]:
response.encoding

'utf-8'

In [15]:
# When you change it, it will be used when you call .text
response.encoding = 'latin-8'
response.text

'[{"id":"20893240420","type":"PullRequestReviewEvent","actor":{"id":29139814,"login":"GCsabi","display_login":"GCsabi","gravatar_id":"","url":"https://api.github.com/users/GCsabi","avatar_url":"https://avatars.githubusercontent.com/u/29139814?"},"repo":{"id":466697561,"name":"joladnijo/joladnijo-webapp","url":"https://api.github.com/repos/joladnijo/joladnijo-webapp"},"payload":{"action":"created","review":{"id":918715493,"node_id":"PRR_kwDOG9E9Wc42wnxl","user":{"login":"GCsabi","id":29139814,"node_id":"MDQ6VXNlcjI5MTM5ODE0","avatar_url":"https://avatars.githubusercontent.com/u/29139814?v=4","gravatar_id":"","url":"https://api.github.com/users/GCsabi","html_url":"https://github.com/GCsabi","followers_url":"https://api.github.com/users/GCsabi/followers","following_url":"https://api.github.com/users/GCsabi/following{/other_user}","gists_url":"https://api.github.com/users/GCsabi/gists{/gist_id}","starred_url":"https://api.github.com/users/GCsabi/starred{/owner}{/repo}","subscriptions_url":

This feature is useful when you have specific knowledge of the encoding. For example, HTML and XML have the ability to specify their encoding in their body. In situations like this, you should use **r.content to find the encoding, and then set r.encoding**. This will let you use r.text with the correct encoding.

## 2.2 Binary Response Content

As the **.text** get you the text version of the response, the **.content** get you the binary version of the response. For example, if you request an image or photo, the response is more likely in byte.

You can use other python module to open properly the byte response. Below example shows how to open an image byte response.

```python
from PIL import Image
from io import BytesIO

i = Image.open(BytesIO(r.content))
```


## 2.3 Json Response Content

There’s also a builtin JSON decoder, in case you’re dealing with JSON data:

In [16]:
r = requests.get('https://api.github.com/events')
r.json()

[{'id': '20895085563',
  'type': 'CreateEvent',
  'actor': {'id': 32468651,
   'login': 'EtienneLt',
   'display_login': 'EtienneLt',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/EtienneLt',
   'avatar_url': 'https://avatars.githubusercontent.com/u/32468651?'},
  'repo': {'id': 105284146,
   'name': 'powsybl/powsybl-core',
   'url': 'https://api.github.com/repos/powsybl/powsybl-core'},
  'payload': {'ref': 'set-loadflow-parameters-with-a-map',
   'ref_type': 'branch',
   'master_branch': 'main',
   'description': 'A framework to build power system oriented software',
   'pusher_type': 'user'},
  'public': True,
  'created_at': '2022-03-23T14:16:58Z',
  'org': {'id': 29916668,
   'login': 'powsybl',
   'gravatar_id': '',
   'url': 'https://api.github.com/orgs/powsybl',
   'avatar_url': 'https://avatars.githubusercontent.com/u/29916668?'}},
 {'id': '20895085519',
  'type': 'PullRequestEvent',
  'actor': {'id': 12514844,
   'login': 'robertkiel',
   'display_login': 'robe

In case the JSON decoding fails, r.json() raises an exception. For example, if the response gets a 204 (No Content), or if the response contains invalid JSON, attempting r.json() raises requests.exceptions.JSONDecodeError. This wrapper exception provides interoperability for multiple exceptions that may be thrown by different python versions and json serialization libraries.

It should be noted that the success of the call to r.json() does not indicate the success of the response. Some servers may return a JSON object in a failed response (e.g. error details with HTTP 500). Such JSON will be decoded and returned. To check that a request is successful, use r.raise_for_status() or check r.status_code is what you expect.

## 2.4 Raw Response Content

In the rare case that you’d like to get the raw socket response from the server, you can access r.raw. If you want to do this, make sure you set stream=True in your initial request. Once you do, you can do this:

In [21]:
r = requests.get('https://api.github.com/events', stream=True)
r.raw

In [24]:
r.raw.read(10)

b''

In general, however, you should write the streamed content to a file like below example:

In [25]:
filename = "/tmp/github_event.txt"
with open(filename, 'wb') as fd:
    for chunk in r.iter_content(chunk_size=128):
        fd.write(chunk)

StreamConsumedError: 

Using Response.iter_content will handle a lot of what you would otherwise have to handle when using Response.raw directly. When streaming a download, the above is the preferred and recommended way to retrieve the content. Note that chunk_size can be freely adjusted to a number that may better fit your use cases.

# 3. Request header

To customize headers, you pass a dictionary of HTTP headers to get() using the **headers parameter**. For example, you can change your previous search request to highlight matching search terms in the results by specifying the text-match media type in the Accept header:

In [27]:
url_param = {'q': 'requests+language:python'}
my_header = {'Accept': 'application/vnd.github.v3.text-match+json'}

response = requests.get(
    'https://api.github.com/search/repositories',
    params=url_param,
    headers=my_header,
)

# View the new `text-matches` array which provides information
# about your search term within the results
json_response = response.json()
repository = json_response['items'][0]
print(f'Text matches: {repository["text_matches"]}')

Text matches: [{'object_url': 'https://api.github.com/repositories/4290214', 'object_type': 'Repository', 'property': 'description', 'fragment': 'Requests + Gevent = <3', 'matches': [{'text': 'Requests', 'indices': [0, 8]}]}]


The Accept header tells the server what content types your application can handle. In this case, since you’re expecting the matching search terms to be highlighted, you’re using the header value application/vnd.github.v3.text-match+json, which is a proprietary GitHub Accept header where the content is a special JSON format.


Note: Custom headers are given less precedence than more specific sources of information. For instance:

- **Authorization headers set with headers= will be overridden** if credentials are specified in .netrc, which in turn will be overridden by the auth= parameter. Requests will search for the netrc file at ~/.netrc, ~/_netrc, or at the path specified by the NETRC environment variable.

- **Authorization headers will be removed** if you get redirected off-host.

- **Proxy-Authorization headers will be overridden** by proxy credentials provided in the URL.

- **Content-Length headers will be overridden** when we can determine the length of the content.

Furthermore, Requests does not change its behavior at all based on which custom headers are specified. The headers are simply passed on into the final request.

Note: All header values must be a string, bytestring, or unicode. While permitted, it’s advised to avoid passing unicode header values.

## 3.1 Authentication header

There are many ways to do authentication in http request. Authentication header is one of them.

### 3.1.1 Api key

In [28]:
api_key = ""
headers = {'Authorization': f'Token {api_key}'}
response = requests.get('https://nautobot.demo.networktocode.com/api/dcim/devices/', headers=headers)

In [29]:
print(response)

<Response [403]>


### 3.1.2 OIDC token



In [31]:
oidc_token = ""
headers = {'Authorization': f'Bearer {oidc_token}'}
base_url = 'https://webexapis.com'
# Get list of rooms
response = requests.get(f'{base_url}/v1/rooms', headers=headers)
print(response)

<Response [401]>


<Response [401]>


### 3.1.3 Custom Token Header
Some APIs require that the API key be provided in a custom header. For example, Cisco Meraki requires that all requests have an X-Cisco-Meraki-API-Key header with the API key as the value.

In [33]:
# The base URI for all requests
base_uri = "https://api.meraki.com/api/v0"

# Set the custom header to include the API key
headers = {'X-Cisco-Meraki-API-Key': api_key}

# Get a list of organizations
response = requests.get(f'{base_uri}/organizations', headers=headers)

print(response)

<Response [404]>


### 3.1.4 Basic authentication

For the basic login with login and password,

In [40]:
from requests.auth import HTTPBasicAuth

login = "admin"
password = "admin"
response = requests.get('https://api.github.com/user', auth=HTTPBasicAuth(login, password))
print(response)

SSLError: HTTPSConnectionPool(host='api.github.com', port=443): Max retries exceeded with url: /user (Caused by SSLError(CertificateError("hostname 'api.github.com' doesn't match either of 'controller.access.network', 'central.access.network', 'wifiaccess.co'")))

## Some example

In [50]:
oidc_token = ""
headers = {'Authorization': f'Bearer {oidc_token}',
           "Content-Type": "application/json",
           "Accept": "application/json"}
url_base = "https://atlas.lab.sspcloud.fr/api/atlas/v2/search"
params = {"attrName": "qualifiedName",
          "attrValuePrefix": "user-pengfei@movies.Character",
          "limit": "10",
          "offset": "0",
          "typeName": "hive_table"}
response = requests.get(f"{url_base}/attribute", params=params, headers=headers)

In [51]:
response.text

'{"queryType":"BASIC","searchParameters":{"typeName":"hive_table","excludeDeletedEntities":false,"includeClassificationAttributes":false,"includeSubTypes":true,"includeSubClassifications":true,"limit":10,"offset":0,"entityFilters":{"attributeName":"qualifiedName","operator":"startsWith","attributeValue":"user-pengfei@movies.Character"}},"entities":[{"typeName":"hive_table","attributes":{"owner":"headless","createTime":1648021108,"qualifiedName":"user-pengfei@movies.Character","name":"Character","description":"favorite stock, test toto"},"guid":"26240d7d-395e-464e-ad58-1efda879d311","status":"ACTIVE","displayText":"Character","classificationNames":[],"meaningNames":[],"meanings":[],"isIncomplete":false,"labels":[]}],"approximateCount":10}'

In [58]:
post_payload={
  "typeName": "hive_table",
  "excludeDeletedEntities": True,
  "entityFilters": {
    "attributeName": "qualifiedName",
    "attributeValue": "user-pengfei@movies.Character",
    "operator": "startsWith"
  },
  "limit": 25,
  "offset": 0,
   "tagFilters": None,
  "attributes": [""]
}
# when your post payload is a json file, don't use data, use json as key argument
response = requests.post(f"{url_base}/basic", json=post_payload, headers=headers)

In [59]:
response.text

'{"queryType":"BASIC","searchParameters":{"typeName":"hive_table","excludeDeletedEntities":true,"includeClassificationAttributes":false,"includeSubTypes":true,"includeSubClassifications":true,"limit":25,"offset":0,"entityFilters":{"attributeName":"qualifiedName","operator":"startsWith","attributeValue":"user-pengfei@movies.Character"},"attributes":[""]},"entities":[{"typeName":"hive_table","attributes":{"owner":"headless","createTime":1648021108,"qualifiedName":"user-pengfei@movies.Character","name":"Character","description":"favorite stock, test toto"},"guid":"26240d7d-395e-464e-ad58-1efda879d311","status":"ACTIVE","displayText":"Character","classificationNames":[],"meaningNames":[],"meanings":[],"isIncomplete":false,"labels":[]}],"approximateCount":10}'

In [ ]:
# data takes string as input type, not dict
response = requests.post(f"{url_base}/basic", data=post_payload, headers=headers)

There are many other point you may want to know about requests

visit this page https://realpython.com/python-requests/#authentication

## Common error response

- 400: Bad Request is the status code to return when the form of the client request is not as the API expects.
- 401: Unauthorized is the status code to return when the client provides no credentials or invalid credentials.
- 403: Forbidden is the status code to return when a client has valid credentials but not enough privileges to perform an action on a resource.
- 404: page not found